Brofian/compilerbau-minijava
# Implement basic version of semantic and type checker

## Additions
- Implemented basic version of semantic and type checker for
-- independent program and class level 
-- Statements: block,return,if,while
-- Expressions: (None, but prepared structure for future implementation and use in statements)
- Added tests for semantic checks
- Included semantic check in main method and cleaning up
- Added specialized SemanticException to throw only in the context of the semantic and type check

## Changes 
These changes were made to the existing AST.scala file in order to make some things work with the semantic check
- Added some type options (NoneType, BoolType, FunctionType)
- Changed Blocks from List[Statement] into new BlockStatement Type
- Added TypedStatement and TypedExpression
- Changed the content of method bodies to a single Statement to conform to the new BlockStatement. The Block itself should be enforced by the lexer already

## What is missing right now
- Many implementations of statement and expression are still not handled 
- Only static TypeAssertions are used currently, while class handling is not implemented and will require another list of an abstracted class tree in the future 
These features will be added in the future, once more elements of the AST are available


## How to test
Check out the feature branch. From there, you can run the tests with `sbt test` or `sbt testonly SemanticChecks` to only run the tests from the SemanticChecks class. This class defines a sample AST and runs some tests over it
